Mistral: add support for structured output (#2633)

snophey opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
Closes #2633 

## Change
I added the glue-code needed to pass a JSON schema from a `ChatRequest` to the Mistral API. I tried to follow established patterns for setting this up, e.g.:

```java
MistralAiChatModel.builder()
                .apiKey(System.getenv("MISTRAL_AI_API_KEY"))
                .modelName(MISTRAL_SMALL_LATEST)
                .temperature(0.1)
                .logRequests(true)
                .logResponses(true)
                .supportedCapabilities(Capability.RESPONSE_FORMAT_JSON_SCHEMA) // this enables structured output
                .build()
```

When no schema is given in the `ChatRequest`, the old behavior of following the format given in model's `responseFormat` is executed.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [X] I have added unit and/or integration tests for my change
- [X] The tests cover both positive and negative cases
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

snophey committed:
Mistral: add support for structured output (#2633)

snophey committed:
Mistral: activate more tests

snophey committed:
Merge branch 'main' into feature/gh-2633-mistral-structured-output

dliubarskyi commented on langchain4j-mistral-ai/src/main/java/dev/langchain4j/model/mistralai/MistralAiChatModel.java:
```diff
@@ -241,11 +259,13 @@ public MistralAiChatModelBuilder modelName(MistralAiChatModelName modelName) {
             return this;
         }
 
+        @Deprecated(forRemoval = true) // TODO move away from this and on to using ResponseFormat type (?)
```
Do you plan to introduce a `responseFormat(ResponseFormat)` method instead?

snophey replied:
I just added `responseFormat(ResponseFormat)` and changed the underlying `responseFormat` type in the chat model. The deprecated methods now map the "old" inputs to a `resourceType` instance. Let me know if this implementation looks okay to you.

dliubarskyi commented on langchain4j-mistral-ai/src/main/java/dev/langchain4j/model/mistralai/MistralAiChatModel.java:
```diff
@@ -352,6 +372,15 @@ public MistralAiChatModelBuilder maxRetries(Integer maxRetries) {
         }
 
+        /**
+         * @param supportedCapabilities
```
this javadoc can be removed as it does not bring any new info

dliubarskyi commented on langchain4j-mistral-ai/src/main/java/dev/langchain4j/model/mistralai/MistralAiChatModel.java:
```diff
@@ -352,6 +372,15 @@ public MistralAiChatModelBuilder maxRetries(Integer maxRetries) {
+         * @param supportedCapabilities
+         * @return {@code this}.
+         */
+        public MistralAiChatModelBuilder supportedCapabilities(Capability... supportedCapabilities) {
```
Please add `supportedCapabilities(Set<Capability>) `as well

dliubarskyi commented on langchain4j-mistral-ai/src/main/java/dev/langchain4j/model/mistralai/MistralAiStreamingChatModel.java:
```diff
@@ -251,11 +270,13 @@ public MistralAiStreamingChatModelBuilder modelName(MistralAiChatModelName model
             return this;
         }
 
+        @Deprecated(forRemoval = true) // TODO move away from this and use ResponseFormat directly (?)
```
same here

dliubarskyi commented on langchain4j-mistral-ai/src/main/java/dev/langchain4j/model/mistralai/MistralAiStreamingChatModel.java:
```diff
@@ -351,6 +372,14 @@ public MistralAiStreamingChatModelBuilder timeout(final Duration timeout) {
         }
 
+        /**
+         * @return {@code this}.
```
same here

dliubarskyi commented on langchain4j-mistral-ai/src/main/java/dev/langchain4j/model/mistralai/MistralAiStreamingChatModel.java:
```diff
@@ -351,6 +372,14 @@ public MistralAiStreamingChatModelBuilder timeout(final Duration timeout) {
+        /**
+         * @return {@code this}.
+         */
+        public MistralAiStreamingChatModelBuilder supportedCapabilities(Capability... supportedCapabilities) {
```
same here (`Set<Capability>`)

dliubarskyi commented on langchain4j-mistral-ai/src/main/java/dev/langchain4j/model/mistralai/internal/api/MistralAiResponseFormat.java:
```diff
@@ -72,8 +88,78 @@ public MistralAiResponseFormatBuilder type(Object type) {
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    @JsonNaming(SnakeCaseStrategy.class)
+    public static class MistralAiJsonSchema {
```
please move it into a separate file

dliubarskyi reviewed (commented):
@snophey thanks a lot!

dliubarskyi commented:
@snophey could you please also update the following documentation pages?
- https://docs.langchain4j.dev/integrations/language-models/
- https://docs.langchain4j.dev/integrations/language-models/mistral-ai

snophey committed:
Mistral: implement review feedback, add more tests

snophey committed:
Mistral: add structured output details to docs

snophey committed:
Merge branch 'main' into feature/gh-2633-mistral-structured-output

snophey commented:
Hi @dliubarskyi , thanks a lot for reviewing my PR. I have addressed your review comments and added some docs. Let me know if further changes are needed.

snophey committed:
Merge branch 'main' into feature/gh-2633-mistral-structured-output

dliubarskyi committed:
Merge branch 'main' into feature/gh-2633-mistral-structured-output

dliubarskyi reviewed (approved):
@snophey thanks a lot!
